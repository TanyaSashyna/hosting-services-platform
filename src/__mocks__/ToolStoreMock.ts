/* eslint-disable @typescript-eslint/no-explicit-any  */
/* eslint-disable @typescript-eslint/no-empty-function  */

import { ToolStore } from "../stores/ToolStore";
import { StatusTool, DnsChangesType } from "../api/ToolApi";

class ToolStoreMock extends ToolStore {
  getToolStatus(): any {}
  addTool(): any {}
  cancelTool(): any {}
  status: StatusTool = "disabled";
  domain = "";
  dnsChangesRequired = true;
  dnsChanges: DnsChangesType[] = [];
  percentage = 0;
  hostReachable = true;
}

const toolStoreMock = new ToolStoreMock();

toolStoreMock.getToolStatus = jest.fn();
toolStoreMock.addTool = jest.fn();
toolStoreMock.cancelTool = jest.fn();

export default toolStoreMock;
