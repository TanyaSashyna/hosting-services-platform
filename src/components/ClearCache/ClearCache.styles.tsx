import styled from "styled-components";
import { Card as _Card, breakpoints } from "@project-name-frontend/ui-package";

export const Card = styled(_Card)`
  .indent-bottom {
    @media (max-width: ${breakpoints.phoneUpperBoundary}) {
      padding-bottom: 0;
    }
  }
`;
