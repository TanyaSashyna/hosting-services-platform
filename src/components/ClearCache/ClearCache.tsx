import { inject, observer } from "mobx-react";
import React, { FC, useState } from "react";
import { CardRow, CardInfo, useMobile, Button } from "@project-name-frontend/ui-package";
import { useNotifications } from "@project-name-frontend/shared-context";

import { Card } from "./ClearCache.styles";
import { ClearCacheModal } from "../ClearCacheModal";
import { ToolStore } from "../../stores/toolStore";
import { getDateFormat } from "../../common/utils";

type InjectedProps = {
  toolStore: ToolStore;
};

type ClearCacheProps = {
  domain: string;
  lastCacheClearedAt: string;
  id: string;
};

export const ClearCache: FC<ClearCacheProps & InjectedProps> = ({
  domain,
  lastCacheClearedAt,
  id,
  toolStore,
}) => {
  const { mobile } = useMobile();
  const { pushToast } = useNotifications();
  const [openClearCacheModal, setOpenClearCacheModal] = useState(false);
  const [loading, setLoading] = useState(false);

  const handleClearCache = async (): Promise<void> => {
    setLoading(true);
    await toolStore
      .clearCache(id)
      .then(() => {
        pushToast({
          title: "Cache cleared",
          description:
            "Visitors should see the most up-to-date content on your site.",
          intent: "positive",
        });
      })
      .catch(() => {
        pushToast({
          title: "Unable to clear cache",
          description: "Please try again.",
          intent: "negative",
        });
      })
      .finally(() => {
        setLoading(false);
        setOpenClearCacheModal(false);
      });
  };

  return (
    <>
      <Card header="Tools" noPad={true}>
        <CardRow border={true}>
          <CardInfo
            content={{
              children: (
                <div>
                  <strong>Edge cache</strong> dictates the duration for
                  which content remains stored in the cache. By default, this
                  period is set to one hour.
                </div>
              ),
            }}
          />
        </CardRow>
        <CardRow
          border={true}
          direction={mobile ? "column" : "row"}
          spacing="space-between"
          align={mobile ? "" : "center"}
        >
          <CardInfo
            header={{
              children: "Clear all cache",
            }}
            content={{
              children:
                "Clearing the entire cache will ensure that your website's visitors experience the most up-to-date version of your site.",
            }}
            details={
              lastCacheClearedAt &&
              `Last cleared: ${getDateFormat(lastCacheClearedAt)}`
            }
            className={mobile ? "indent-bottom" : ""}
          />
          <CardInfo
            align={mobile ? "center" : "right"}
            action={{
              fullWidth: true,
              children: (
                <Button
                  fullWidth={mobile}
                  onClick={(): void => {
                    setOpenClearCacheModal(true);
                  }}
                  data-testid="clear-cache-btn"
                >
                  Clear all cache
                </Button>
              ),
            }}
          />
        </CardRow>
      </Card>
      <ClearCacheModal
        domain={domain}
        loading={loading}
        open={openClearCacheModal}
        onClose={setOpenClearCacheModal}
        handleClearCache={handleClearCache}
      />
    </>
  );
};

ClearCache.displayName = "ClearCache";

export default inject("toolStore")(
  observer(ClearCache)
) as unknown as FC<ClearCacheProps>;
