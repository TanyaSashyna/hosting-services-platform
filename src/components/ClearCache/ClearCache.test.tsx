import React from "react";
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import { ClearCache } from "./ClearCache";
import ToolStoreMock from "../../__mocks__/ToolStoreMock";

afterEach(cleanup);

describe("ClearCache", () => {
  it("renders ClearCache and matches snapshot", () => {
    const { container, getByText, asFragment } = render(
      <ClearCache
        domain="example.com"
        id="id12345"
        lastCacheClearedAt=""
        toolStore={ToolStoreMock}
      />
    );

    expect(container).toBeDefined();
    expect(getByText("Tools")).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders Last cleared", () => {
    const mockDate = "2023-10-23T09:47:31.472Z";
    const { container } = render(
      <ClearCache
        domain="example.com"
        id="id12345"
        lastCacheClearedAt={mockDate}
        toolStore={ToolStoreMock}
      />
    );

    expect(container.querySelector(".cardInfo__details")).toBeDefined();
  });

  it("ropen the ClearCacheModal", () => {
    const mockDate = "2023-10-23T09:47:31.472Z";
    const { getByText, getByTestId } = render(
      <ClearCache
        domain="example.com"
        id="id12345"
        lastCacheClearedAt={mockDate}
        toolStore={ToolStoreMock}
      />
    );

    const clearCacheBtn = getByTestId("clear-cache-btn");
    fireEvent.click(clearCacheBtn);

    expect(
      waitFor(() => expect(getByText("Clear all cache for example.com")))
    ).toBeDefined();
  });
});
