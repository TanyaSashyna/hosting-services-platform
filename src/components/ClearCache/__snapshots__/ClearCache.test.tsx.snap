// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearCache renders ClearCache and matches snapshot 1`] = `
<DocumentFragment>
  @media (max-width:768px) {
  .c0 .indent-bottom {
    padding-bottom: 0;
  }
}

<div
    class="card c0"
  >
    <div
      class="card__head"
    >
      <div
        class="card__head__text"
      >
        <div
          class="card__head--wrap"
        >
          <div
            class="card__title"
          >
            Tools
          </div>
        </div>
      </div>
    </div>
    <div
      class="card__body card__body--no-pad"
    >
      <div
        class="cardRow cardRow--border"
      >
        <div
          class="cardInfo"
        >
          <div
            class="cardInfo__inner"
          >
            <span
              class="cardInfo__content"
            >
              <div>
                <strong>
                  Edge cache
                </strong>
                 dictates the duration for which content remains stored in the cache. By default, this period is set to one hour.
              </div>
            </span>
          </div>
        </div>
      </div>
      <div
        class="cardRow cardRow--direction-column cardRow--spacing-space-between cardRow--border"
      >
        <div
          class="cardInfo indent-bottom"
        >
          <div
            class="cardInfo__inner"
          >
            <span
              class="cardInfo__header"
            >
              Clear all cache
            </span>
            <span
              class="cardInfo__content"
            >
              Clearing the entire cache will ensure that your website's visitors experience the most up-to-date version of your site.
            </span>
          </div>
        </div>
        <div
          class="cardInfo cardInfo--align-center"
        >
          <div
            class="cardInfo__inner cardInfo--fullWidth"
          >
            <div
              class="cardInfo__action"
            >
              <button
                class="button button--size-default button--type-default button--text button--fullWidth"
                data-testid="clear-cache-btn"
                type="button"
              >
                <div
                  class="button__internal"
                >
                  <div
                    class="button__children"
                  >
                    Clear all cache
                  </div>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
